### PR TITLE
Strip epoch on packages that use it in pre-cache

### PR DESCRIPTION
**Pull request by dmcilvaney**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [ ] The toolchain has been rebuilt successfully (or no changes were made to it)
- [ ] The toolchain/worker package manifests are up-to-date
- [ ] Any updated packages successfully build (or no packages were changed)
- [ ] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [ ] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [ ] All package sources are available
- [ ] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [ ] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [ ] All source files have up-to-date hashes in the `*.signatures.json` files
- [ ] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [ ] Documentation has been updated to match any changes to the build system
- [ ] Ready to merge

---

###### Summary 

Based on the contents of rpms-snapshot the pre-cache tool was calculating versions of the form `1:2.3.4-5`, of note is the epoch value `1`. The tool uses these values to compute the expected .rpm filepaths, but these paths do not contain the epoch number. Now we try to patch an rpm both with and without the epoch. We try both so we can explicitly warn the user we will not download the version with epoch (since `:` is a special character in `Make` and we use the contents of the rpm cache as an input into the rest of the tools).

###### Change Log  

- String `<epoch>:` from the version of packages being checked by the pre-cache tool.

###### Does this affect the toolchain?  

**NO**

###### Test Methodology

- Local

### CODE REVIEW

File: toolkit/tools/precacher/precacher.go [hunk 1]
```diff
@@ -160,8 +162,11 @@ func getAllRepoData(repoURLs []string, workerTar, buildDir string) (namesToURLs
 			// We need to prepend the repoURL to the partial URL to get the full URL
 			packageRepoPath = fmt.Sprintf("%s/%s", repoURL, packageRepoPath)
 			namesToURLs[packageName] = packageRepoPath
+			URLList = append(URLList, packageRepoPath)
 		}
 	}
+	file.WriteLines(URLList, listFile)
```
Comment (PawelWMS) [inline toolkit/tools/precacher/precacher.go:168]:
```suggestion
	err = file.WriteLines(URLList, listFile)
```

File: toolkit/tools/precacher/precacher.go [hunk 2]
```diff
@@ -59,6 +59,7 @@ var (
 	outDir            = app.Flag("output-dir", "Directory to download packages into.").Required().ExistingDir()
 	snapshot          = app.Flag("snapshot", "Path to the rpm snapshot .json file.").ExistingFile()
 	outputSummaryFile = app.Flag("output-summary-file", "Path to save the summary of packages downloaded").String()
+	repoSummaryFile   = app.Flag("repo-summary-file", "Path to save the list of packages available in the repos").String()
```
Comment (PawelWMS) [inline toolkit/tools/precacher/precacher.go:62]: Can you rephrase to clarify this is the list of URLs to the packages, not just package names themselves?

File: toolkit/scripts/precache.mk [hunk 3]
```diff
@@ -34,6 +35,7 @@ $(STATUS_FLAGS_DIR)/precache.flag: $(go-precacher) $(chroot_worker) $(rpms_snaps
 		--snapshot "$(rpms_snapshot)" \
 		--output-dir "$(remote_rpms_cache_dir)" \
 		--output-summary-file "$(precache_downloaded_files)" \
+		--repo-summary-file "$(repo_summary_file)" \
```
Comment (PawelWMS) [inline toolkit/scripts/precache.mk:38]: Small request: can we use a different word than "summary-file". We've been using that word for the JSON files containing the contents of the `RepoContents` structs. Best to avoid overloading terms, if possible.

File: toolkit/tools/precacher/precacher.go [hunk 4]
```diff
@@ -350,12 +355,12 @@ func precachePackage(pkg *repocloner.RepoPackage, packagesAvailableFromRepos map
 	const (
 		downloadRetryAttempts = 2
 		downloadRetryDuration = time.Second
+		stripEpoch            = false
+		leaveEpoch            = true
```
Comment (PawelWMS) [inline toolkit/tools/precacher/precacher.go:359]: Nit-pick: alphabetical order.
```suggestion
		leaveEpoch            = true
		stripEpoch            = false
```

Comment (PawelWMS) [inline toolkit/tools/precacher/precacher.go:359]: File names for RPM packages never contain the epoch number - that's only encoded in the RPMs metadata inside these files. We can just always strip the epoch from the version, if it's present and simplify the code a bit.

File: toolkit/tools/precacher/precacher.go (hunk 1 above)
Comment (dmcilvaney) [reply toolkit/tools/precacher/precacher.go:168]: Done

File: toolkit/tools/precacher/precacher.go (hunk 4 above)
Comment (dmcilvaney) [reply toolkit/tools/precacher/precacher.go:359]: removed

Comment (dmcilvaney) [reply toolkit/tools/precacher/precacher.go:359]: Good to know, I've simplified this then, and just always strip the epoch if its present in the version

File: toolkit/tools/precacher/precacher.go [hunk 5]
```diff
@@ -377,7 +381,7 @@ func precachePackage(pkg *repocloner.RepoPackage, packagesAvailableFromRepos map
 		return
 	}
 
-	// Get the url for the package, or bail out if it is not available
+	// Get the url for the package, or bail out if it is not available.
```
Comment (PawelWMS) [inline toolkit/tools/precacher/precacher.go:384]: Since we're here...:)
```suggestion
	// Get the URL for the package, or bail out if it is not available.
```

**Review by PawelWMS — APPROVED** [review]
